File: python/agent/minimal-agent-example/minimal_agent.py
import os
import json
import warnings
from galileo import log, galileo_context, openai as galileo_openai
from dotenv import load_dotenv
from rich.console import Console
import questionary
import openai

# Suppress Pydantic serializer warnings
warnings.filterwarnings("ignore", message="Pydantic serializer warnings")

load_dotenv()
console = Console()
client = galileo_openai.OpenAI(api_key=os.environ.get("OPENAI_API_KEY"))
openai_client = openai.OpenAI(api_key=os.environ.get("OPENAI_API_KEY"))


# Tool: Convert text numbers to numerical values
@log(span_type="tool", name="convert_text_to_arithmetic_expression")
def convert_text_to_arithmetic_expression(text):
    """Convert a text number (like 'seven') to its numerical value (7)."""
    console.print(f"Converting: {text}")

    response = openai_client.chat.completions.create(
        model="gpt-4o",
        messages=[
            {
                "role": "user",
                "content": f"Convert the expression {text} with text numbers to a valid arithmetic expression. Respond with only the arithmetic expression.",
            }
        ],
    )

    result = response.choices[0].message.content.strip()
    try:
        expression = result
        console.print(f"Converted to: {expression}")
        return str(expression)
    except:
        console.print(f"Error parsing: {result}")
        return "Error"


# Tool: Calculator for arithmetic operations
@log(span_type="tool", name="calculate")
def calculate(expression):
    """Perform a calculation based on the given expression."""
    console.print(f"Calculating: {expression}")

    try:
        result = eval(expression)
        console.print(f"Result: {result}")
        return f"The result of {expression} is {result}"
    except Exception as e:
        return f"Error calculating {expression}: {str(e)}"


# Load tools from tools.json
def get_tools():
    with open(os.path.join(os.path.dirname(__file__), "tools.json"), "r") as f:
        return json.load(f)


# Main processing function
@log(span_type="llm")
def process_query(query):
    """Process a numerical query using LLM and tools."""
    console.print(f"Processing query: {query}")

    # Initialize conversation history with serializable dictionaries
    messages = [
        {
            "role": "system",
            "content": "Convert text expressions to valid arithmetic expressions, then calculate results. Use tools one at a time.",
        },
        {"role": "user", "content": f"Process this query: '{query}'"},
    ]

    tools = get_tools()
    results = []

    # Agent loop - continue until the LLM decides we're done
    while True:
        # Get next tool call from LLM
        response = client.chat.completions.create(
            model="gpt-4o",
            messages=messages,
            tools=tools,
        )

        # Convert the assistant message to a serializable dictionary
        assistant_message = response.choices[0].message
        assistant_dict = {"role": "assistant", "content": assistant_message.content}

        # Add tool calls if present
        if assistant_message.tool_calls:
            tool_calls_list = []
            for tool_call in assistant_message.tool_calls:
                tool_calls_list.append(
                    {
                        "id": tool_call.id,
                        "type": "function",
                        "function": {
                            "name": tool_call.function.name,
                            "arguments": tool_call.function.arguments,
                        },
                    }
                )
            assistant_dict["tool_calls"] = tool_calls_list

        messages.append(assistant_dict)  # Add assistant's response to history

        # Check if LLM is done (no more tool calls)
        if not assistant_message.tool_calls:
            # LLM provided a final answer
            if assistant_message.content:
                results.append(assistant_message.content)
            break

        # Process the tool call
        for tool_call in assistant_message.tool_calls:
            function_name = tool_call.function.name
            function_args = json.loads(tool_call.function.arguments)

            console.print(f"\n[bold]Executing {function_name} tool:[/bold]")

            # Execute the appropriate tool
            if function_name == "convert_text_to_arithmetic_expression":
                text = function_args["text"]
                result = convert_text_to_arithmetic_expression(text)
            elif function_name == "calculate":
                expression = function_args["expression"]
                result = calculate(expression)
            else:
                result = f"Unknown tool: {function_name}"

            # Add the tool result to conversation history as a serializable dictionary
            tool_result_message = {
                "tool_call_id": tool_call.id,
                "role": "tool",
                "name": function_name,
                "content": result,
            }
            messages.append(tool_result_message)
            results.append(result)

    # Create a summary
    if results:
        summary = "\n".join(results)
    else:
        summary = "No results produced."

    return summary


def main():
    console.print("[bold]Minimal Number Converter & Calculator[/bold]")

    query = questionary.text("Enter your query:", default="What's 4 + seven?").ask()

    if query is None or query.lower() in ["exit", "quit", "q"]:
        console.print("Exiting.")
        return

    # Process the query within a Galileo context
    with galileo_context():
        result = process_query(query)

    console.print("\n[bold green]Result:[/bold green]")
    console.print(result)


if __name__ == "__main__":
    try:
        main()
    except KeyboardInterrupt:
        console.print("\nExiting.")

File: typescript/agent/minimal-agent-example/minimal_agent.ts
import * as dotenv from 'dotenv';
import * as fs from 'fs';
import * as path from 'path';
import { OpenAI } from 'openai';
import { log, wrapOpenAI, init, flush } from 'galileo';
import chalk from 'chalk';
import inquirer from 'inquirer';

// Load environment variables
dotenv.config();

// Suppress any warnings if needed
// process.env.NODE_TLS_REJECT_UNAUTHORIZED = '0';

// Get Galileo configuration from environment variables
const projectName = process.env.GALILEO_PROJECT || 'agent_test_typescript';
const logStreamName = process.env.GALILEO_LOG_STREAM || 'dev';

// Initialize clients
const client = wrapOpenAI(new OpenAI({ apiKey: process.env.OPENAI_API_KEY }));
const openaiClient = new OpenAI({ apiKey: process.env.OPENAI_API_KEY });

// Tool: Convert text to arithmetic expression
const convertTextToArithmeticExpression = log(
  { spanType: 'tool', name: 'convert_text_to_arithmetic_expression' },
  async (text: string): Promise<string> => {
    console.log(chalk.blue(`Converting: ${text}`));
    
    const response = await openaiClient.chat.completions.create({
      model: "gpt-4o",
      messages: [
        { 
          role: "user", 
          content: `Convert the expression ${text} with text numbers to a valid arithmetic expression. Respond with only the arithmetic expression.`
        }
      ],
    });
    
    const result = response.choices[0].message.content?.trim() || '';
    try {
      const expression = result;
      console.log(chalk.green(`Converted to: ${expression}`));
      return expression;
    } catch (error) {
      console.log(chalk.red(`Error parsing: ${result}`));
      return "Error";
    }
  }
);

// Tool: Calculator for arithmetic operations
const calculate = log(
  { spanType: 'tool', name: 'calculate' },
  async (expression: string): Promise<string> => {
    console.log(chalk.blue(`Calculating: ${expression}`));
    
    try {
      // Using Function instead of eval for slightly better security
      // Still not recommended for production without proper validation
      const result = Function(`'use strict'; return (${expression})`)();
      console.log(chalk.green(`Result: ${result}`));
      return `The result of ${expression} is ${result}`;
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error);
      return `Error calculating ${expression}: ${errorMessage}`;
    }
  }
);

// Load tools from tools.json
function getTools(): any[] {
  try {
    const toolsPath = path.join(__dirname, 'tools.json');
    const toolsData = fs.readFileSync(toolsPath, 'utf8');
    return JSON.parse(toolsData);
  } catch (error) {
    console.error('Error loading tools:', error);
    return [];
  }
}

// Main processing function
const processQuery = log(
  { spanType: 'llm' },
  async (query: string): Promise<string> => {
    console.log(chalk.blue(`Processing query: ${query}`));
    
    // Initialize conversation history with serializable dictionaries
    const messages: any[] = [
      { role: "system", content: "Convert text expressions to valid arithmetic expressions, then calculate results. Use tools one at a time." },
      { role: "user", content: `Process this query: '${query}'` }
    ];
    
    const tools = getTools();
    const results: string[] = [];
    
    // Agent loop - continue until the LLM decides we're done
    while (true) {
      // Get next tool call from LLM
      const response = await client.chat.completions.create({
        model: "gpt-4o",
        messages: messages,
        tools: tools,
      });
      
      // Convert the assistant message to a serializable dictionary
      const assistantMessage = response.choices[0].message;
      const assistantDict: any = {
        role: "assistant",
        content: assistantMessage.content
      };
      
      // Add tool calls if present
      if (assistantMessage.tool_calls && assistantMessage.tool_calls.length > 0) {
        const toolCallsList = assistantMessage.tool_calls.map(toolCall => ({
          id: toolCall.id,
          type: "function",
          function: {
            name: toolCall.function.name,
            arguments: toolCall.function.arguments
          }
        }));
        assistantDict.tool_calls = toolCallsList;
      }
          
      messages.push(assistantDict);  // Add assistant's response to history
      
      // Check if LLM is done (no more tool calls)
      if (!assistantMessage.tool_calls || assistantMessage.tool_calls.length === 0) {
        // LLM provided a final answer
        if (assistantMessage.content) {
          results.push(assistantMessage.content);
        }
        break;
      }
      
      // Process the tool call
      for (const toolCall of assistantMessage.tool_calls || []) {
        const functionName = toolCall.function.name;
        const functionArgs = JSON.parse(toolCall.function.arguments);
        
        console.log(chalk.bold(`\nExecuting ${functionName} tool:`));
        
        // Execute the appropriate tool
        let result: string;
        if (functionName === "convert_text_to_arithmetic_expression") {
          const text = functionArgs.text;
          result = await convertTextToArithmeticExpression(text);
        } else if (functionName === "calculate") {
          const expression = functionArgs.expression;
          result = await calculate(expression);
        } else {
          result = `Unknown tool: ${functionName}`;
        }
        
        // Add the tool result to conversation history as a serializable dictionary
        const toolResultMessage = {
          tool_call_id: toolCall.id,
          role: "tool",
          name: functionName,
          content: result
        };
        messages.push(toolResultMessage);
        results.push(result);
      }
    }
    
    // Create a summary
    if (results.length > 0) {        
      return results.join('\n');
    } else {
      return "No results produced.";
    }
  }
);

async function main() {
  console.log(chalk.bold('Minimal Number Converter & Calculator'));
  
  // Initialize Galileo with project and log stream names
  init({
    projectName,
    logStreamName
  });
  
  console.log(chalk.green(`✅ Project: ${projectName}`));
  console.log(chalk.green(`✅ Log Stream: ${logStreamName}`));
  
  try {
    const { query } = await inquirer.prompt([
      {
        type: 'input',
        name: 'query',
        message: 'Enter your query:',
        default: "What's 4 + seven?"
      }
    ]);
    
    if (!query || ['exit', 'quit', 'q'].includes(query.toLowerCase())) {
      console.log('Exiting.');
      return;
    }
    
    // Process the query
    const result = await processQuery(query);
    
    console.log(chalk.bold.green('\nResult:'));
    console.log(result);
  } catch (error) {
    console.error('Error:', error);
  } finally {
    // Flush Galileo logs
    await flush();
  }
}

// Run the main function
if (require.main === module) {
  main().catch(error => {
    console.error('Unhandled error:', error);
    process.exit(1);
  });
} 